internal/cli/cmd/cluster/grpcproxy.go

```diff
@@ -0,0 +1,208 @@
+	p, err := newGrpcProxy(connect)
+	if err != nil {
+		return err
+	}
+
+	return p.Serve(listener)
+}
+
+type grpcProxy struct {
+	connect func(context.Context) (net.Conn, error)
+	*grpc.Server
+
+	mu sync.Mutex
+	// Fields protected by mutex go below
+	backendClient *grpc.ClientConn
+}
+
+func newGrpcProxy(connect func(context.Context) (net.Conn, error)) (*grpcProxy, error) {
+	g := &grpcProxy{
+		connect: connect,
+	}
+
+	g.Server = grpc.NewServer(grpc.UnknownServiceHandler(g.handler))
+	return g, nil
+}
+
+func (g *grpcProxy) newBackendClient(ctx context.Context) (*grpc.ClientConn, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.backendClient != nil {
+		if g.backendClient.GetState() == connectivity.Ready {
+			fmt.Fprintf(console.Debug(ctx), "reused grpc connection\n")
+			return g.backendClient, nil
+		}
+
+		fmt.Fprintf(console.Debug(ctx), "cached grpc connection invalidated: %v\n", g.backendClient.GetState())
+		g.backendClient = nil
+	}
+
+	client, err := grpc.DialContext(ctx, "",
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    time.Second * 30,
+			Timeout: time.Minute,
+		}),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
+			return g.connect(ctx)
+		}))
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Fprintf(console.Debug(ctx), "created new grpc connection\n")
+
+	g.backendClient = client
+	return client, nil
+}
+
+func (g *grpcProxy) handler(srv interface{}, serverStream grpc.ServerStream) error {
+	ctx := serverStream.Context()
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		err := status.Errorf(codes.Internal, "reading method failed")
+		fmt.Fprintf(console.Debug(ctx), "reading method failed: %v\n", err)
+		return err
+	}
+
+	fmt.Fprintf(console.Debug(ctx), "handler %s\n", fullMethodName)
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	outgoingCtx := metadata.NewOutgoingContext(serverStream.Context(), md.Copy())
+	backendConn, err := g.newBackendClient(outgoingCtx)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "creating backend connection failed: %v\n", err)
+		return status.Errorf(codes.Internal, "failed to connect to backend: %v", err)
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "failed to create client stream: %v\n", err)
+		return status.Errorf(codes.Internal, "failed create client: %v", err)
+	}
+
+	s2cErrChan := g.proxyServerToClient(serverStream, clientStream)
+	c2sErrChan := g.proxyClientToServer(clientStream, serverStream)
+	// Make sure to close both client and server connections
+	for i := 0; i < 2; i++ {
+		select {
+		case s2cErr := <-s2cErrChan:
```

Comment:
if the channel is closed, s2cErr will be read as `nil` -- I guess that's ok?

If detecting a closed channel is important, then you can do `case s2cErr, ok := <-s2ErrChan:`

Reply:
The problem with closed channels is that they don't block the `select`, so I need to set them to `nil`. See new commit.